lib/rgeo/shapefile/reader.rb

```diff
@@ -340,12 +342,20 @@ def next
 
       # Read the remaining records starting with the current record index,
       # and yield the Reader::Record for each one.
-
       def each
-        return unless @opened
-        yield _read_next_record while @cur_record_index < @num_records
+        return to_enum(:each) { @num_records } unless block_given?
+        raise "File was not open" unless @opened
+        # Each needs to be idempotent, therefore we reset all the internal indexes to their original value
+        current_record_index = @cur_record_index
+        rewind
```

Comment:
Your begin block is much more clear than my ensure indeed, however, I guess this `rewind` should be part of it